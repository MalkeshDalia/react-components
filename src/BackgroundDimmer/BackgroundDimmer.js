import React from 'react';
import ReactDOM from 'react-dom';
import cx from 'classnames';
import { props, t, skinnable } from '../utils';
import FlexView from 'react-flexview';
import omit from 'lodash/omit';

export const Props = {
  children: t.ReactChildren,
  color: t.maybe(t.String),
  alpha: t.maybe(t.Number),
  zIndex: t.maybe(t.Number),
  stopScrollPropagation: t.maybe(t.Boolean),
  onClickOutside: t.maybe(t.Function),
  width: t.maybe(t.union([t.String, t.Number])),
  maxWidth: t.maybe(t.union([t.String, t.Number])),
  height: t.maybe(t.union([t.String, t.Number])),
  maxHeight: t.maybe(t.union([t.String, t.Number])),
  className: t.maybe(t.String),
  id: t.maybe(t.String),
  style: t.maybe(t.Object)
};

/**
 * Creates a full-page dimmed background for its children nodes
 * @param children - children nodes/elements
 * @param color - background-color
 * @param alpha - opacity
 * @param zIndex - z-index (BackgroundDimmer has `position: fixed`)
 * @param stopScrollPropagation - avoid propagation for scroll events
 * @param onClickOutside - called when user clicks outside children
 * @param width - centeredContentWrapper width
 * @param maxWidth - centeredContentWrapper max-width
 * @param height - centeredContentWrapper height
 * @param maxHeight - centeredContentWrapper max-height
 */
@props(Props)
@skinnable()
export default class BackgroundDimmer extends React.Component {

  static defaultProps = {
    color: 'black',
    alpha: 0.5,
    zIndex: 99999,
    width: 'auto',
    maxWidth: '90%',
    height: 'auto',
    maxHeight: '90%'
  };

  isEventOutsideChildren = (e) => {
    const el = e.target || e.srcElement;
    return el === ReactDOM.findDOMNode(this.refs.mainContentWrapper);
  };

  onClick = (e) => {
    const { onClickOutside } = this.props;
    if (onClickOutside) {
      onClickOutside(e);
    }
  };

  stopPropagation = e => e.stopPropagation();

  preventDefault = e => e.preventDefault();

  stopScrollPropagation = (e) => {
    if (this.props.stopScrollPropagation && this.isEventOutsideChildren(e)) {
      this.preventDefault(e);
    }
  };

  getLocals() {
    const {
      onClick, stopPropagation, stopScrollPropagation,
      props: { className, zIndex, color, alpha, width, maxWidth, height, maxHeight, ...props }
    } = this;

    const fixedStyle = { position: 'fixed', top: 0, left: 0, right: 0, bottom: 0 };
    return {
      ...omit(props, ['onClickOutside']),
      className: cx('background-dimmer', className),
      overlayProps: {
        style: {
          ...fixedStyle,
          zIndex,
          backgroundColor: color,
          opacity: String(alpha)
        }
      },
      mainContentWrapperProps: {
        onClick,
        onWheel: stopScrollPropagation,
        onTouchMove: stopScrollPropagation,
        style: { ...fixedStyle, zIndex: (zIndex + 1) },
        className: 'main-content-wrapper',
        vAlignContent: 'center',
        hAlignContent: 'center',
        ref: 'mainContentWrapper'
      },
      centeredContentWrapperProps: {
        className: 'centered-content-wrapper',
        style: { width, maxWidth, height, maxHeight },
        onClick: stopPropagation,
        column: true
      }
    };
  }

  template({ children, overlayProps, mainContentWrapperProps, centeredContentWrapperProps, ...locals }) {
    return (
      <div {...locals}>
        <div {...overlayProps} />
        <FlexView {...mainContentWrapperProps}>
          <FlexView {...centeredContentWrapperProps}>
            {children}
          </FlexView>
        </FlexView>
      </div>
    );
  }

}
