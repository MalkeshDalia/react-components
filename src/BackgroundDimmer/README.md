# BackgroundDimmer

Creates a full-page dimmed background for its children nodes

## Props
|Name|Type|Default|Description|
|----|----|-------|-----------|
| **children** | <code>ReactChildren</code> |  | **required**. Children nodes/elements |
| **color** | <code>String</code> | <code>"black"</code> | *optional*. Background-color |
| **alpha** | <code>Number</code> | <code>0.5</code> | *optional*. Opacity |
| **zIndex** | <code>Number</code> | <code>99999</code> | *optional*. Z-index (BackgroundDimmer has `position: fixed`) |
| **stopScrollPropagation** | <code>Boolean</code> |  | *optional*. Avoid propagation for scroll events |
| **onClickOutside** | <code>Function</code> |  | *optional*. Called when user clicks outside children |
| **width** | <code>union(String &#124; Number)</code> | <code>"auto"</code> | *optional*. CenteredContentWrapper width |
| **maxWidth** | <code>union(String &#124; Number)</code> | <code>"90%"</code> | *optional*. CenteredContentWrapper max-width |
| **height** | <code>union(String &#124; Number)</code> | <code>"auto"</code> | *optional*. CenteredContentWrapper height |
| **maxHeight** | <code>union(String &#124; Number)</code> | <code>"90%"</code> | *optional*. CenteredContentWrapper max-height |
| **className** | <code>String</code> |  | *optional*. Additional `className` for wrapper element |
| **id** | <code>String</code> |  | *optional*. Custom `id` for wrapper element |
| **style** | <code>Object</code> |  | *optional*. Inline-style overrides for wrapper element |