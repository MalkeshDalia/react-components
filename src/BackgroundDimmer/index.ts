import BackgroundDimmer, { BackgroundDimmerProps, Props } from './BackgroundDimmer';
export default BackgroundDimmer;
export { BackgroundDimmerProps, Props };
