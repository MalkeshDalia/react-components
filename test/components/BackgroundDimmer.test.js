import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import BackgroundDimmer from '../../src/BackgroundDimmer';

const content = <div className='content'>content</div>;

describe('BackgroundDimmer', () => {

  it('renders correctly', () => {
    const component = renderer.create(
      <BackgroundDimmer>
        {content}
      </BackgroundDimmer>
    );
    expect(component).toMatchSnapshot();
  });

  it('invokes the callback when the background is clicked', () => {
    const onClickOutside = jest.fn();
    const dimmer = shallow(
      <BackgroundDimmer onClickOutside={onClickOutside}>
        {content}
      </BackgroundDimmer>
    );
    dimmer.find('.main-content-wrapper').simulate('click');
    expect(onClickOutside).toBeCalled();
  });

  it('doesn\'t invoke the callback when the content is clicked', () => {
    const onClickOutside = jest.fn();
    const dimmer = shallow(
      <BackgroundDimmer onClickOutside={onClickOutside}>
        {content}
      </BackgroundDimmer>
    );
    dimmer.find('.content').simulate('click');
    expect(onClickOutside).not.toBeCalled();
  });

  describe('getLocals', () => {

    it('uses default props', () => {
      const dimmer = new BackgroundDimmer({
        ...BackgroundDimmer.defaultProps,
        children: content
      });
      const locals = dimmer.getLocals();
      expect(locals).toMatchSnapshot();
    });

  });

});
