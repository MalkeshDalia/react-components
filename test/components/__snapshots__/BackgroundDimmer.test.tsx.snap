// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundDimmer renders correctly 1`] = `
<div
  className="background-dimmer"
>
  <div
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "left": 0,
        "opacity": 0.5,
        "position": "fixed",
        "right": 0,
        "top": 0,
        "zIndex": 99999,
      }
    }
  />
  <div
    className="react-flex-view align-content-center justify-content-center main-content-wrapper"
    onClick={[Function]}
    onTouchMove={[Function]}
    onWheel={[Function]}
    style={
      Object {
        "MozBoxFlex": "0 1 auto",
        "WebkitBoxFlex": "0 1 auto",
        "WebkitFlex": "0 1 auto",
        "bottom": 0,
        "flex": "0 1 auto",
        "left": 0,
        "msFlex": "0 1 auto",
        "position": "fixed",
        "right": 0,
        "top": 0,
        "zIndex": 100000,
      }
    }
  >
    <div
      className="react-flex-view flex-column centered-content-wrapper"
      onClick={[Function]}
      style={
        Object {
          "MozBoxFlex": "0 1 auto",
          "WebkitBoxFlex": "0 1 auto",
          "WebkitFlex": "0 1 auto",
          "flex": "0 1 auto",
          "height": "auto",
          "maxHeight": "90%",
          "maxWidth": "90%",
          "msFlex": "0 1 auto",
          "width": "auto",
        }
      }
    >
      <div
        className="content"
      >
        content
      </div>
    </div>
  </div>
</div>
`;
